lib/types.nix

```diff
@@ -352,6 +352,25 @@ rec {
       functor = (defaultFunctor name) // { wrapped = [ t1 t2 ]; };
     };
 
+    coercedTo = coercedType: coerceFunc: finalType: mkOptionType rec {
+      name = "coercedTo";
+      description = "${finalType.description} or ${coercedType.description} coerced to it";
+      check = x: coercedType.check x || finalType.check x;
+      merge = loc: defs:
+        let
+          coerceVal = val: if coercedType.check val
```

Comment:
This code will always attempt to coerce.  If there is an overlap between the types, I am not sure this would be the expected behavior. Checking for `finalType.check` before `coercedType.check` would make more sense to prevent coercing data which already have the final type.